File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        // create postfix
        int postfix = 1;
        int[] res = new int[nums.length];
        res[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            // res[0] = 1
            // res[1] = (res[0]=1) * (nums[0]=1) = 1
            // res [2] = (res[1]=1) * (nums[1]=2) = 2
            // res [3] = (res[2]=2) * (nums[2]=4) = 8
            res[i] = res[i - 1] * nums[i - 1];
        }

        for (int i = nums.length - 1; i >= 0; i--) {
            // res[3] = 8 * 1
            // postfix = 6
            res[i] = res[i] * postfix;
            postfix = postfix * nums[i];
        }
        return res;
    }
}  

File: Data Structures & Algorithms/sliding-window-maximum/submission-0.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # Brute force solution
        # for every index I will find the maximum value in the window
        res = []
        # need to figure out when it'll stop though
        # so if len(nums) = 7
        # last window has to contain 5,6,7 without going out of bounds
        # so we need to iterate all the way up to 5
        # let len(nums) = m 
        # so the general case will be range = m - k + 1 
        # that will give us 7 - 3 + 1 = 5
        for i in range(0, len(nums) - k + 1):
            # window start is at i and end is at k
            window = nums[i: i+k]
            # get the max element of the window
            max_element = max(window)
            # insert max element into array
            res.append(max_element)

        return res